group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/GeneralUniformAssignmentBuilder.java

```diff
@@ -14,17 +14,891 @@
+            if (memberData == null) {
+                LOG.warn("Member Id {} not found", memberId);
+                return 0;
+            }
+            return memberData.currentAssignmentSize;
+        }
+
+        /**
+         * @param memberId      The member Id.
+         * @return The maximum assignment size for the given member.
+         */
+        public int maxAssignmentSize(String memberId) {
+            MemberAssignmentData memberData = this.membersWithAssignmentSizes.get(memberId);
+            if (memberData == null) {
+                LOG.warn("Member Id {} not found", memberId);
+                return 0;
+            }
+            return memberData.maxAssignmentSize;
+        }
+
+        /**
+         * @param memberId      The member Id.
+         * Increment the current target assignment size for the member.
+         */
+        public void incrementTargetAssignmentSize(String memberId) {
+            MemberAssignmentData memberData = this.membersWithAssignmentSizes.get(memberId);
+            if (memberData == null) {
+                LOG.warn("Member Id {} not found", memberId);
+                return;
+            }
+            memberData.currentAssignmentSize++;
+        }
+
+        /**
+         * @param memberId      The member Id.
+         * Decrement the current target assignment size for the member, if it's assignment size is greater than zero.
+         */
+        public void decrementTargetAssignmentSize(String memberId) {
+            MemberAssignmentData memberData = this.membersWithAssignmentSizes.get(memberId);
+            if (memberData == null) {
+                LOG.warn("Member Id {} not found", memberId);
+                return;
+            }
+
+            if (memberData.currentAssignmentSize > 0) {
+                memberData.currentAssignmentSize --;
+            }
+        }
+
+        /**
+         * Computes and sets the maximum assignment size for each member.
+         */
+        private void initializeAssignmentSizes() {
+            members.forEach((memberId, member) -> {
+                int maxSize = member.subscribedTopicIds().stream()
+                    .mapToInt(subscribedTopicDescriber::numPartitions)
+                    .sum();
+
+                MemberAssignmentData memberAssignmentData = membersWithAssignmentSizes
+                    .computeIfAbsent(memberId, MemberAssignmentData::new);
+                memberAssignmentData.maxAssignmentSize = maxSize;
+                memberAssignmentData.currentAssignmentSize = 0;
+            });
+        }
+
+        /**
+         * Assigns partition to member if eligible.
+         *
+         * @param topicIdPartition      The partition to be assigned.
+         * @param memberId              The Id of the member.
+         * @return true if the partition was assigned, false otherwise.
+         */
+        protected boolean maybeAssignPartitionToMember(
```

Comment:
I have noticed that we use have both `public` and `protected` methods in this class and I am not sure to understand why. Could you help me to understand this?

Reply:
this is my thought process do correct me if I'm wrong, I used public and protected so that the outer class can access the methods. B/w public and protected I used public for methods that could potentially be reused by other classes that need this assignment manager functionality but I realize this access is limited by the protected nature of the class itself. 
I think for now I can make the class private and all the methods package private so that they can be accessed by the outer class? The access modifiers always confuse me, I'm open to suggestions in this matter! Thanks!

Reply:
Yeah, keeping it internal for now makes sense to me.